File: mobile-app-flutter/lib/features/products/services/product_service.dart
import '../../../core/network/api_client.dart';
import '../models/product_model.dart';

class ProductService {
  final ApiClient _apiClient;

  ProductService(this._apiClient);

  Future<Product> getProduct(String id) async {
    final response = await _apiClient.get('/products/$id');
    return Product.fromJson(response.data['product']);
  }

  Future<List<Product>> getFeaturedProducts({int limit = 10}) async {
    final response = await _apiClient.get('/products', queryParameters: {
      'featured': true,
      'limit': limit,
    });
    return (response.data['products'] as List)
        .map((p) => Product.fromJson(p))
        .toList();
  }

  Future<List<Product>> getNewProducts({int limit = 10}) async {
    final response = await _apiClient.get('/products', queryParameters: {
      'sort': 'created_at',
      'order': 'desc',
      'limit': limit,
    });
    return (response.data['products'] as List)
        .map((p) => Product.fromJson(p))
        .toList();
  }

  Future<List<Product>> getPopularProducts({int limit = 10}) async {
    final response = await _apiClient.get('/products', queryParameters: {
      'sort': 'rating',
      'order': 'desc',
      'limit': limit,
    });
    return (response.data['products'] as List)
        .map((p) => Product.fromJson(p))
        .toList();
  }

  Future<List<Product>> getProductsByCategory(String categoryId, {int page = 1, int limit = 20}) async {
    final response = await _apiClient.get('/products', queryParameters: {
      'category_id': categoryId,
      'page': page,
      'limit': limit,
    });
    return (response.data['products'] as List)
        .map((p) => Product.fromJson(p))
        .toList();
  }

  Future<List<Product>> searchProducts(String query, {
    int page = 1,
    int limit = 20,
    String? categoryId,
    double? minPrice,
    double? maxPrice,
    String? sortBy,
    String? sortOrder,
  }) async {
    final response = await _apiClient.get('/products/search', queryParameters: {
      'q': query,
      'page': page,
      'limit': limit,
      if (categoryId != null) 'category_id': categoryId,
      if (minPrice != null) 'min_price': minPrice,
      if (maxPrice != null) 'max_price': maxPrice,
      if (sortBy != null) 'sort': sortBy,
      if (sortOrder != null) 'order': sortOrder,
    });
    return (response.data['products'] as List)
        .map((p) => Product.fromJson(p))
        .toList();
  }

  Future<List<Product>> getRelatedProducts(String productId, {int limit = 6}) async {
    final response = await _apiClient.get('/products/$productId/related', queryParameters: {
      'limit': limit,
    });
    return (response.data['products'] as List)
        .map((p) => Product.fromJson(p))
        .toList();
  }
}


  Future<Product> getProductById(String id) async {
    final response = await _apiClient.get('/products/$id');
    return Product.fromJson(response.data['product']);
  }

  Future<List<Product>> getNewArrivals({int limit = 10}) async {
    final response = await _apiClient.get('/products', queryParameters: {
      'sort': 'created_at',
      'order': 'desc',
      'limit': limit,
    });
    return (response.data['products'] as List).map((p) => Product.fromJson(p)).toList();
  }

  Future<List<Product>> getBestSellers({int limit = 10}) async {
    final response = await _apiClient.get('/products', queryParameters: {
      'sort': 'sales',
      'order': 'desc',
      'limit': limit,
    });
    return (response.data['products'] as List).map((p) => Product.fromJson(p)).toList();
  }

File: archive/legacy-services/wholesale-service/src/services/product.service.ts
// Product Service - Wholesale B2B
// خدمة المنتجات - البيع بالجملة

import { PrismaClient, WholesaleProduct, ProductStatus } from '@prisma/client';
import { v4 as uuidv4 } from 'uuid';

const prisma = new PrismaClient();

interface CreateProductDTO {
  supplierId: string;
  name: string;
  nameAr?: string;
  description: string;
  descriptionAr?: string;
  category: string;
  subcategory?: string;
  brand?: string;
  basePrice: number;
  currency?: string;
  stockQuantity: number;
  minOrderQuantity?: number;
  maxOrderQuantity?: number;
  unitType?: string;
  unitsPerPackage?: number;
  weight?: number;
  dimensions?: { length: number; width: number; height: number };
  specifications?: Record<string, any>;
  images: string[];
  videos?: string[];
  documents?: string[];
  countryOfOrigin?: string;
  hsCode?: string;
  certifications?: string[];
}

interface UpdateProductDTO {
  name?: string;
  nameAr?: string;
  description?: string;
  descriptionAr?: string;
  basePrice?: number;
  stockQuantity?: number;
  minOrderQuantity?: number;
  maxOrderQuantity?: number;
  images?: string[];
  status?: ProductStatus;
}

interface ProductFilters {
  supplierId?: string;
  category?: string;
  subcategory?: string;
  countryOfOrigin?: string;
  minPrice?: number;
  maxPrice?: number;
  inStock?: boolean;
  status?: ProductStatus;
}

export class ProductService {
  // ==========================================
  // 📦 PRODUCT MANAGEMENT
  // ==========================================

  // Create product
  async createProduct(data: CreateProductDTO): Promise<WholesaleProduct> {
    const sku = `WS-${data.supplierId.slice(0, 4).toUpperCase()}-${uuidv4().slice(0, 8).toUpperCase()}`;

    return prisma.wholesaleProduct.create({
      data: {
        ...data,
        sku,
        currency: data.currency || 'USD',
        minOrderQuantity: data.minOrderQuantity || 1,
        unitType: data.unitType || 'piece',
        unitsPerPackage: data.unitsPerPackage || 1,
        videos: data.videos || [],
        documents: data.documents || [],
        certifications: data.certifications || [],
        status: 'DRAFT'
      }
    });
  }

  // Get product by ID
  async getProductById(id: string): Promise<WholesaleProduct | null> {
    return prisma.wholesaleProduct.findUnique({
      where: { id },
      include: {
        supplier: {
          select: {
            id: true,
            businessName: true,
            businessNameAr: true,
            country: true,
            rating: true,
            isVerified: true
          }
        },
        pricingTiers: {
          orderBy: { minQuantity: 'asc' }
        }
      }
    });
  }

  // Update product
  async updateProduct(id: string, data: UpdateProductDTO): Promise<WholesaleProduct> {
    return prisma.wholesaleProduct.update({
      where: { id },
      data
    });
  }

  // Delete product (soft delete)
  async deleteProduct(id: string): Promise<WholesaleProduct> {
    return prisma.wholesaleProduct.update({
      where: { id },
      data: {
        status: 'DISCONTINUED',
        isActive: false
      }
    });
  }

  // List products with filters
  async listProducts(
    filters: ProductFilters,
    page: number = 1,
    limit: number = 20,
    sortBy: string = 'createdAt',
    sortOrder: 'asc' | 'desc' = 'desc'
  ): Promise<{ products: WholesaleProduct[]; total: number; pages: number }> {
    const where: any = { isActive: true };

    if (filters.supplierId) where.supplierId = filters.supplierId;
    if (filters.category) where.category = filters.category;
    if (filters.subcategory) where.subcategory = filters.subcategory;
    if (filters.countryOfOrigin) where.countryOfOrigin = filters.countryOfOrigin;
    if (filters.status) where.status = filters.status;
    if (filters.inStock) where.stockQuantity = { gt: 0 };
    if (filters.minPrice || filters.maxPrice) {
      where.basePrice = {};
      if (filters.minPrice) where.basePrice.gte = filters.minPrice;
      if (filters.maxPrice) where.basePrice.lte = filters.maxPrice;
    }

    const [products, total] = await Promise.all([
      prisma.wholesaleProduct.findMany({
        where,
        skip: (page - 1) * limit,
        take: limit,
        orderBy: { [sortBy]: sortOrder },
        include: {
          supplier: {
            select: {
              id: true,
              businessName: true,
              country: true,
              rating: true,
              isVerified: true
            }
          },
          pricingTiers: true
        }
      }),
      prisma.wholesaleProduct.count({ where })
    ]);

    return {
      products,
      total,
      pages: Math.ceil(total / limit)
    };
  }

  // ==========================================
  // 💰 PRICING TIERS
  // ==========================================

  // Add pricing tier
  async addPricingTier(
    productId: string,
    minQuantity: number,
    maxQuantity: number | null,
    pricePerUnit: number
  ): Promise<any> {
    return prisma.productPricingTier.create({
      data: {
        productId,
        minQuantity,
        maxQuantity,
        pricePerUnit
      }
    });
  }

  // Get price for quantity
  async getPriceForQuantity(productId: string, quantity: number): Promise<{
    unitPrice: number;
    totalPrice: number;
    savings: number;
    tier: string;
  }> {
    const product = await prisma.wholesaleProduct.findUnique({
      where: { id: productId },
      include: {
        pricingTiers: {
          orderBy: { minQuantity: 'desc' }
        }
      }
    });

    if (!product) throw new Error('Product not found');

    // Find applicable tier
    let unitPrice = product.basePrice;
    let tier = 'Base';

    for (const pricingTier of product.pricingTiers) {
      if (quantity >= pricingTier.minQuantity) {
        if (!pricingTier.maxQuantity || quantity <= pricingTier.maxQuantity) {
          unitPrice = pricingTier.pricePerUnit;
          tier = `${pricingTier.minQuantity}+ units`;
          break;
        }
      }
    }

    const totalPrice = unitPrice * quantity;
    const baseTotalPrice = product.basePrice * quantity;
    const savings = baseTotalPrice - totalPrice;

    return { unitPrice, totalPrice, savings, tier };
  }

  // ==========================================
  // 📊 INVENTORY
  // ==========================================

  // Update stock
  async updateStock(id: string, quantity: number, operation: 'add' | 'subtract' | 'set'): Promise<WholesaleProduct> {
    const product = await prisma.wholesaleProduct.findUnique({ where: { id } });
    if (!product) throw new Error('Product not found');

    let newQuantity: number;
    switch (operation) {
      case 'add':
        newQuantity = product.stockQuantity + quantity;
        break;
      case 'subtract':
        newQuantity = Math.max(0, product.stockQuantity - quantity);
        break;
      case 'set':
        newQuantity = quantity;
        break;
    }

    const status = newQuantity === 0 ? 'OUT_OF_STOCK' : product.status === 'OUT_OF_STOCK' ? 'ACTIVE' : product.status;

    return prisma.wholesaleProduct.update({
      where: { id },
      data: {
        stockQuantity: newQuantity,
        status
      }
    });
  }

  // Check availability
  async checkAvailability(id: string, quantity: number): Promise<{
    available: boolean;
    currentStock: number;
    minOrder: number;
    maxOrder: number | null;
  }> {
    const product = await prisma.wholesaleProduct.findUnique({ where: { id } });
    if (!product) throw new Error('Product not found');

    return {
      available: product.stockQuantity >= quantity && quantity >= product.minOrderQuantity,
      currentStock: product.stockQuantity,
      minOrder: product.minOrderQuantity,
      maxOrder: product.maxOrderQuantity
    };
  }

  // ==========================================
  // 🔍 SEARCH
  // ==========================================

  // Search products
  async searchProducts(
    query: string,
    filters: ProductFilters,
    page: number = 1,
    limit: number = 20
  ): Promise<{ products: WholesaleProduct[]; total: number }> {
    const where: any = {
      isActive: true,
      status: 'ACTIVE',
      OR: [
        { name: { contains: query, mode: 'insensitive' } },
        { nameAr: { contains: query, mode: 'insensitive' } },
        { description: { contains: query, mode: 'insensitive' } },
        { sku: { contains: query, mode: 'insensitive' } },
        { brand: { contains: query, mode: 'insensitive' } }
      ]
    };

    if (filters.category) where.category = filters.category;
    if (filters.countryOfOrigin) where.countryOfOrigin = filters.countryOfOrigin;

    const [products, total] = await Promise.all([
      prisma.wholesaleProduct.findMany({
        where,
        skip: (page - 1) * limit,
        take: limit,
        include: {
          supplier: {
            select: {
              id: true,
              businessName: true,
              country: true,
              rating: true
            }
          }
        }
      }),
      prisma.wholesaleProduct.count({ where })
    ]);

    return { products, total };
  }

  // Get categories
  async getCategories(): Promise<{ category: string; count: number }[]> {
    const categories = await prisma.wholesaleProduct.groupBy({
      by: ['category'],
      where: { isActive: true, status: 'ACTIVE' },
      _count: true
    });

    return categories.map(c => ({
      category: c.category,
      count: c._count
    }));
  }
}

export const productService = new ProductService();
